Particle Provider API

Leo40Git opened the pull request:
Closes #93.

This API mirrors Fabric's equivalent API, meaning you can easily create animated particles that work on both sides!

Leo40Git committed:
Particle Factory API (closes architectury#93)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/architectury/architectury-api?pullRequest=110) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/architectury/architectury-api?pullRequest=110) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/architectury/architectury-api?pullRequest=110) it.</sub>

Leo40Git committed:
Add testmod

Added public static modifiers to methods (OOPS)

Leo40Git commented:
I've added a test to the testmod (summon the test particle via `/particle` - I tried adding a block that randomly creates that particle but that failed).

Also, tested on Fabric 1.16.5.

Leo40Git committed:
Remove unused accessor

shedaniel commented on common/src/main/java/me/shedaniel/architectury/registry/ParticleFactories.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @FunctionalInterface
+    public interface PendingParticleProvider<T extends ParticleOptions> {
```
DeferredParticleProvider

shedaniel commented on common/src/main/java/me/shedaniel/architectury/registry/ParticleFactories.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @ExpectPlatform
+    public static <T extends ParticleOptions> void register(ParticleType<T> type, PendingParticleProvider<T> constructor) {
```
Param: constructor -> provider
(fix this for the subprojects as well)

Leo40Git committed:
Requested changes

PendingParticleProvider -> DeferredParticleProvider
parameter constructor -> provider

MaxNeedsSnacks commented:
Adding last call since Dan‘s review has been resolved

shedaniel committed:
Some changes

- Migrate from ParticleFactories to ParticleProviderRegistry (We are not yarn)
- Format Code

Signed-off-by: shedaniel <daniel@shedaniel.me>
